File: packages/sdk-python/veltrix_sdk/resources/connectivity.py
# -*- coding: utf-8 -*-

from .base import BaseResource

class ConnectivityResource(BaseResource):
    """Handles operations related to component connectivity details."""
    RESOURCE_PATH = "connectivity"

    def get_for_component(self, component_id, **kwargs):
        """
        Retrieves connectivity details for a specific component.

        Args:
            component_id (str): The ID of the component.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The connectivity details object for the component.
        """
        # Corresponds to GET /api/connectivity/component/{componentId}
        path = f"{self.RESOURCE_PATH}/component/{component_id}"
        return self._http_client.get(path, **kwargs)

    def update_for_component(self, component_id, status=None, ssh_command=None, https_url=None, tailscale_key=None, tailscale_device_id=None, tailscale_device_ip=None, **kwargs):
        """
        Updates connectivity details for a specific component.

        Args:
            component_id (str): The ID of the component.
            status (str, optional): New status.
            ssh_command (str, optional): New SSH command.
            https_url (str, optional): New HTTPS URL.
            tailscale_key (str, optional): New Tailscale key.
            tailscale_device_id (str, optional): New Tailscale device ID.
            tailscale_device_ip (str, optional): New Tailscale device IP.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The updated connectivity details object.
        """
        # Corresponds to PUT /api/connectivity/component/{componentId}
        path = f"{self.RESOURCE_PATH}/component/{component_id}"
        data = {
            "status": status,
            "sshCommand": ssh_command,
            "httpsUrl": https_url,
            "tailscaleKey": tailscale_key,
            "tailscaleDeviceId": tailscale_device_id,
            "tailscaleDeviceIP": tailscale_device_ip,
        }
        data = {k: v for k, v in data.items() if v is not None}
        return self._http_client.put(path, data=data, **kwargs)

    def delete_for_component(self, component_id, **kwargs):
        """
        Deletes connectivity details for a specific component.

        Args:
            component_id (str): The ID of the component.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: A confirmation message (API returns 200 OK with message).
        """
        # Corresponds to DELETE /api/connectivity/component/{componentId}
        # API spec indicates 200 OK with message
        path = f"{self.RESOURCE_PATH}/component/{component_id}"
        return self._http_client.delete(path, **kwargs)

    def create_or_update(self, component_id, status=None, ssh_command=None, https_url=None, **kwargs):
        """
        Creates or updates connectivity details for a component.

        Args:
            component_id (str): The ID of the component.
            status (str, optional): Status of the connectivity.
            ssh_command (str, optional): SSH command for connection.
            https_url (str, optional): HTTPS URL for connection.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The created or updated connectivity details object.
        """
        # Corresponds to POST /api/connectivity/
        # Note: This uses the base path POST, not the component-specific path
        data = {
            "componentId": component_id,
            "status": status,
            "sshCommand": ssh_command,
            "httpsUrl": https_url,
        }
        data = {k: v for k, v in data.items() if v is not None}
        return self._create(data=data, **kwargs)  # Using base _create method

    def regenerate_key(self, component_id, **kwargs):
        """
        Regenerates the connectivity key for a specific component.

        Args:
            component_id (str): The ID of the component.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The updated connectivity details object with the new key.
        """
        # Corresponds to POST /api/connectivity/component/{componentId}/regenerate-key
        path = f"{self.RESOURCE_PATH}/component/{component_id}/regenerate-key"
        # Use http_client directly as it's a POST to a sub-path with no body
        return self._http_client.post(path, **kwargs)

File: packages/sdk-js/src/resources/connectivity.ts
import { BaseResource } from './base';
import { AxiosRequestConfig } from 'axios';
import { APIError } from '../errors';

// Define interfaces for Connectivity data structures
interface ConnectivityDetails {
  id: string; // uuid
  componentId: string; // uuid
  status?: string | null;
  sshCommand?: string | null;
  httpsUrl?: string | null;
  tailscaleKey?: string | null; // Usually masked or omitted on GET
  tailscaleDeviceId?: string | null;
  tailscaleDeviceIP?: string | null;
  createdAt: string; // ISO Date string
  updatedAt: string; // ISO Date string
}

interface CreateOrUpdateConnectivityPayload {
  componentId: string; // uuid
  status?: string | null;
  sshCommand?: string | null;
  httpsUrl?: string | null;
}

interface UpdateConnectivityPayload {
  status?: string | null;
  sshCommand?: string | null;
  httpsUrl?: string | null;
  tailscaleKey?: string | null;
  tailscaleDeviceId?: string | null;
  tailscaleDeviceIP?: string | null;
}

export class ConnectivityResource extends BaseResource {
  protected readonly RESOURCE_PATH = "connectivity";

  async getForComponent(componentId: string, config?: AxiosRequestConfig): Promise<ConnectivityDetails> {
    // Corresponds to GET /api/connectivity/component/{componentId}
    const path = `${this.RESOURCE_PATH}/component/${componentId}`;
    return this.httpClient.get<ConnectivityDetails>(path, config);
  }

  async updateForComponent(componentId: string, payload: UpdateConnectivityPayload, config?: AxiosRequestConfig): Promise<ConnectivityDetails> {
    // Corresponds to PUT /api/connectivity/component/{componentId}
    const path = `${this.RESOURCE_PATH}/component/${componentId}`;
    const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this.httpClient.put<ConnectivityDetails>(path, cleanedPayload, config);
  }

  async deleteForComponent(componentId: string, config?: AxiosRequestConfig): Promise<{ message: string }> {
    // Corresponds to DELETE /api/connectivity/component/{componentId}
    // API Spec indicates 200 OK with message
    const path = `${this.RESOURCE_PATH}/component/${componentId}`;
    // Use http_client directly as base delete expects 204
     const result = await this.httpClient.delete<{ message: string } | null>(path, config);
     if (result === null) {
         throw new APIError(`DELETE ${path} returned null/204, expected 200 with message.`);
     }
     return result;
  }

  async createOrUpdate(payload: CreateOrUpdateConnectivityPayload, config?: AxiosRequestConfig): Promise<ConnectivityDetails> {
    // Corresponds to POST /api/connectivity/
    const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    // Uses base path POST
    return this._create<ConnectivityDetails>(cleanedPayload, undefined, config);
  }

  async regenerateKey(componentId: string, config?: AxiosRequestConfig): Promise<ConnectivityDetails> {
    // Corresponds to POST /api/connectivity/component/{componentId}/regenerate-key
    const path = `${this.RESOURCE_PATH}/component/${componentId}/regenerate-key`;
    // Use http_client directly as it's a POST to a sub-path with no body
    return this.httpClient.post<ConnectivityDetails>(path, undefined, config);
  }
}
